### ISSUE TITLE
best practice to create UI to enable/dsiable render of a Layer ( Tileset ) and add 3D model to map

### ISSUE THREAD

**Issue by Maverick87Shaka** [mapbox/mapbox-unity-sdk#534]
Support ticket:
* Unity version:  2017.3.0f3
* Mapbox SDK version: MapBox 1.3.0 Release 
* The platform you're building to: Windows Steam VR - HTC Vive and other HMD
* A description of what you're trying to do: 
I would like to create an app where it would be possible to select in the UI witch Layer to render or not over the map.  I Follow the tutorial and on each example the layer is "hardcoded" inside the "map" creation. Actually inside my studio I have a StyleMap including all my Layer ( Tilesets ). If I use in Unity the custom map ID, I have inside Unity all the tile with the Layer as "texture" so there is now way to turn on or off.
I also try to take a look to the TerrainVectorMap example, that seams to be closer to what I'm looking for, but unfortunately I can't manage to enable/disable the "building" layer at runtime. Witch is the best way to do this?

I'm also looking on how to place a custom full 3D object over the map with high precision and matching the scale on "dynamic zoom" example. Something like placing a full 3D modeled bridge across the river.

Thanks!

**Comment by david-rhodes** [comment]
I recommend starting with layers and camera culling. For example, you can add a `LayerModifier` to your building stack (with layer `Buildings`). Then you can toggle the building layer in the camera culling options. Alternatively, just register your buildings to a manager object which can toggle the gameobjects at will.

**Comment by brnkhy** [comment]
`For example, you can add a LayerModifier to your building stack (with layer Buildings). Then you can toggle the building layer in the camera culling options.`
This is a beautiful solution, I like it.

`Alternatively, just register your buildings to a manager object which can toggle the gameobjects at will.`
This was my first thought. Main difference from the other one would be that hidden objects will be disabled in this solution and monobehaviours on them won't run/update etc. We're also planning to release such manager objects (object collections) soon to simplify this kind of solutions.

**Comment by Maverick87Shaka** [comment]
I tried the LayerModifier and it works great. In a couple of minutes I write the UI to update the culling mask of my main camera to turn on/off a specific layer! 
Thanks you all for your suggestion, I'm looking forward to the new release to improve behaviors and performance, for now I'm satisfied of the result with the LayerModifier.

Now I have a "little" doubt about mapbox studio, since I can't create a single tileset with multiple layer, I need to create a different tileset for each layer, this is forcing me to create multiple VectorTileFactory and add all of them to my map, instead of creating one single VectorTileFactory and create multiple VectorLayerVisualizer, is that correct?

**Comment by brnkhy** [comment]
@Maverick87Shaka 
` I can't create a single tileset with multiple layer, I need to create a different tileset for each layer, this is forcing me to create multiple VectorTileFactory and add all of them to my map, instead of creating one single VectorTileFactory and create multiple VectorLayerVisualizer, is that correct?`

MapIds support multiple tilesets through a comma separated input. So even if you have 10 tilesets, each can have a layer and you can use a map id like `mapbox.mapbox-streets-v7,mapbox.mapbox-terrain-v2,mapbox.mapbox-traffic-v1` in one vector tile factory to fetch them all.
When you do that, server actually merges all those tileset into one response so you'll get them all in one single server request. Having multiple vector factories will also work but that'll do a separate request for each factory.

**Comment by Maverick87Shaka** [comment]
Switched from multiple vector factories to a single vector factories with multiple MAP ID comma separated works great! Thanks for tip @brnkhy !

**Comment by brnkhy** [comment]
@Maverick87Shaka awesome!  I'm closing this for now, feel free to open it again or create a new issue if you have follow up questions.

**Comment by Maverick87Shaka** [comment]
@brnkhy Adding more layer I encounter a what I think is related to "system/filesystem" limitation.
When adding more then 7 MAP ID in my case, the mapbox layer are not visualized anymore, due to a problem fetching the cache file. I notice that using 7 tilesets the filename is about 180 characters, when adding eighth tileset ( MAP ID ) the cache filename reach 205 characters, and I suppose this trigger some exception in the SQLite or system due to limitation also on filesystem.

My system is Windows 10 Enterprise and Filesystem is NTFS. 

Note: the filename cache is depending on username, in my case is maverick87shaka, so with a "shorter" username the number of tilesets could be higher before triggering this error.

`SQLiteException: Could not open database file: C:/Users/sacob0xh/AppData/LocalLow/DefaultCompany/MapBox-SDK1_3\cache\mapbox.mapbox-streets-v7,maverick87shaka.ccs35n1h,maverick87shaka.2yd1ucj1,maverick87shaka.7pz5puin,maverick87shaka.b5cmjr1r,maverick87shaka.bbg43nz9,maverick87shaka.91mjuilb,maverick87shaka.1xdch9a4.cache (CannotOpen)
SQLite4Unity3d.SQLiteConnection..ctor (System.String databasePath, SQLiteOpenFlags openFlags, Boolean storeDateTimeAsTicks) (at Assets/Mapbox/Core/mapbox-sdk-cs/Platform/SQLite/SQLite.cs:206)
Mapbox.Platform.MbTiles.MbTilesDb.openOrCreateDb (System.String dbName) (at Assets/Mapbox/Core/mapbox-sdk-cs/Platform/MbTiles/MbTiles.cs:132)
Mapbox.Platform.MbTiles.MbTilesDb..ctor (System.String tileset, Nullable`1 maxTileCount) (at Assets/Mapbox/Core/mapbox-sdk-cs/Platform/MbTiles/MbTiles.cs:32)
Mapbox.Platform.Cache.MbTilesCache.initializeMbTiles (System.String mapId) (at Assets/Mapbox/Core/mapbox-sdk-cs/Platform/Cache/MbTilesCache.cs:101)
Mapbox.Platform.Cache.MbTilesCache.Get (System.String mapId, CanonicalTileId tileId) (at Assets/Mapbox/Core/mapbox-sdk-cs/Platform/Cache/MbTilesCache.cs:124)
Mapbox.Platform.Cache.CachingWebFileSource.Request (System.String uri, System.Action`1 callback, Int32 timeout, CanonicalTileId tileId, System.String mapId) (at Assets/Mapbox/Core/mapbox-sdk-cs/Platform/Cache/CachingWebFileSource.cs:112)
Mapbox.Unity.MapboxAccess.Request (System.String url, System.Action`1 callback, Int32 timeout, CanonicalTileId tileId, System.String mapId) (at Assets/Mapbox/Unity/MapboxAccess.cs:171)
Mapbox.Map.Tile.Initialize (IFileSource fileSource, CanonicalTileId canonicalTileId, System.String mapId, System.Action p) (at Assets/Mapbox/Core/mapbox-sdk-cs/Map/Tile.cs:137)
Mapbox.Unity.MeshGeneration.Factories.VectorTileFactory.OnRegistered (Mapbox.Unity.MeshGeneration.Data.UnityTile tile) (at Assets/Mapbox/Unity/MeshGeneration/Factories/VectorTileFactory.cs:90)
Mapbox.Unity.MeshGeneration.Factories.AbstractTileFactory.Register (Mapbox.Unity.MeshGeneration.Data.UnityTile tile) (at Assets/Mapbox/Unity/MeshGeneration/Factories/AbstractTileFactory.cs:86)
Mapbox.Unity.Map.AbstractMapVisualizer.LoadTile (UnwrappedTileId tileId) (at Assets/Mapbox/Unity/Map/AbstractMapVisualizer.cs:193)
Mapbox.Unity.Map.AbstractMap.TileProvider_OnTileAdded (UnwrappedTileId tileId) (at Assets/Mapbox/Unity/Map/AbstractMap.cs:184)
Mapbox.Unity.Map.AbstractTileProvider.AddTile (UnwrappedTileId tile) (at Assets/Mapbox/Unity/Map/AbstractTileProvider.cs:34)
Mapbox.Unity.Map.QuadTreeTileProvider.Update () (at Assets/Mapbox/Unity/Map/QuadTreeTileProvider.cs:146)`

![mapbox4](https://user-images.githubusercontent.com/51521/36199978-8555b74c-117b-11e8-9ff1-68f543f3d04e.PNG)

**Comment by brnkhy** [comment]
@Maverick87Shaka aah it makes sense. Yes there's a path string length limit in windows so it get even worse when you project is burried in some other folders etc. 
@BergWerkGIS would probably know more about this.

**Comment by BergWerkGIS** [comment]
@Maverick87Shaka @brnkhy 
Yes this is a long standing known issue of Windows: its internal `MAX_PATH` is fixed to **260 characters**.

One question @Maverick87Shaka : 
Do you really always need all the data of +7 tilesets?
Depending on what they contain they might get big and slow down your app unnecessarily.
You could create a style referencing all your tilesets and only the layers needed and then get the tiles via our style optimized API.

Nevertheless there are several ways to work around the `MAX_PATH` issue:

1. Prefix the file name with `\\?\` eg
```
\\?\C:\my\very\long\path
```

Could you try to prefix [`_dbPath` here](https://github.com/mapbox/mapbox-unity-sdk/blob/1898596df8727255628a830e136c4fa8e4f4bb2d/sdkproject/Assets/Mapbox/Core/mapbox-sdk-cs/Platform/MbTiles/MbTiles.cs#L129) with `\\?\` ?

**Please let me know if this works for you then I'll get it into the SDK.**


2. Fiddle with the registry.

3. Group Policy Editor: `Filesystem -> Enable Win32 long paths`



`2.` and `3.` are described here: 
https://www.howtogeek.com/266621/how-to-make-windows-10-accept-file-paths-over-260-characters/

Their caveat of course being that those steps would have to be taken on each machine that encounters the problem.
So I'm clearly in favor of `1.` but if it doesn't work one of the others should definitely do the trick for you.

**Comment by Maverick87Shaka** [comment]
Hi @BergWerkGIS ,
I don't really know if I need to use all the tilesets individually, but it's seams the common way to be able to turn on/off the render of specific layer. Using Style in my map with all the layer generated from tilesets result in a map with layer already "merged" in the texture map , and I don't know how to call update of the map to show/hide specific layer. If there is a better way to do achieve this behavior I'll use it for sure ;)

My line now it's `_dbPath = Path.Combine(Application.persistentDataPath, "cache");`, the default of the SDK. On my system it' results in: `C:/Users/zzzzzzzz/AppData/LocalLow/DefaultCompany/MapBox-SDK1.3\cache` . If I set manually the _`dbPath = "C:\\tmp";` can I use more Tilesets without trigger the error. Could be a temporary workaround at the moment for me. 

How can I set the prefix `\\?\` on the Path.Combine?
I also try to set this prefix manually in the string path but this creates new exception. 
`_dbPath = "\\?\C:\\tmp";`
![image](https://user-images.githubusercontent.com/51521/36665076-e0cf3980-1ae6-11e8-9ed7-8640079f3891.png)

**Comment by BergWerkGIS** [comment]
@Maverick87Shaka 

> I don't really know if I need to use all the tilesets individually, but it's seams the common way to be able to turn on/off the render of specific layer. Using Style in my map with all the layer generated from tilesets result in a map with layer already "merged" in the texture map , and I don't know how to call update of the map to show/hide specific layer. If there is a better way to do achieve this behavior I'll use it for sure ;)

As long as you don't run into any performance issues with your current approach I think you are good.
Otherwise you might want to look into the style optimized API:
https://blog.mapbox.com/style-optimized-vector-tiles-39868da81275

----

> How can I set the prefix \\?\ on the Path.Combine?

1. as your error message hints you have to escape the blackslashes, either like this
  ```c#
  "\\\\?\\"
  ```
  or like that
  ```c#
  @"\\?\"
  ```

2. when using `\\?\` you also have to get rid of the forward slashes returned by `Application.persistentDataPath`.
Either doing or manual `Replace()` or letting .Net take care of it, eg via `Path.GetFullPath`

So replacing this line

https://github.com/mapbox/mapbox-unity-sdk/blob/b539a41851bcd8f3c08db626bcf0546913b5ef5e/sdkproject/Assets/Mapbox/Core/mapbox-sdk-cs/Platform/MbTiles/MbTiles.cs#L142

with

```c#
_dbPath = @"\\?\" + Path.Combine(Application.persistentDataPath, "cache");
_dbPath = Path.GetFullPath(_dbPath);
```

should do the trick.

**Comment by Maverick87Shaka** [comment]
Thanks, I've now adjusted the script, but unfortunately there is some more error. On my system I've enabled the support for long path, and if I try to create manually the namefile in that specific cache path with a program like OneCommanderV2 it's works, is not working trying to create the same filename with Explorer, but it's seams a common limitation of explorer. Maybe Unity or sqlite is facing the same limitation.  

For now i'm solving my personal issue forcing c:\tmp\ as cache path to save some characters.
